Add support for scipy.basinhopping

reneeotten opened the pull request:
As discussed in #440, including some more global optimization algorithms wouldn't hurt. This PR adds support for the ```basinhopping``` algorithm as implemented in ```scipy```. 

It passes bounds to the underlying optimizer (i.e., the default optimizer is now ```L-BFGS-B``` since we're using bounds) and it uses the ```accept_test``` callable to accept only steps that stay within the parameter bounds.

In addition, one test is included to make sure that the implementation in ```lmfit``` gives the same results as directly calling ```scipy.optimize.basinhopping```. And just because I noticed, I also update the name of the example for the ```brute``` method.

reneeotten commented:
@newville okay, I rebased this after merging the previous PR. The last two commits make sure the ```AbortFitException``` is handled correctly and I merged the ```penalty``` and ```penalty_brute``` functions into one (the only difference is the ```apply_bounds_transformation```.

reneeotten commented:
well... it appears that the argument ```seed``` was added in ```scipy``` version 0.19, that's why some of the Travis tests fail. What is the preference here, just bump the minimum required version or should I work around this (i.e., remove the argument when a lower version of ```scipy``` is used)?

newville commented:
@reneeotten  I think the last time "minimum scipy version requirement" came up, we changed the requirement to support a version that was still 2 years old at that time.   I think that if you asked me without looking at version history, I would say that 2 years is about right.   But given the ease and prevalence of virtual environments and conda, and given the release of scipy 1.0,  making that "1 year" or even require scipy 1.0 seems fine, especially if there are changes relevant to lmfit. 

It turns out that scipy 0.19 is just over 1 year old.   For "2 years old", we'd still have to support 0.17.   Adding tests for older versions that we'll just want to remove it anyway is kind of ugly. 

So I would say that requiring scipy 0.19 for the next release is fine, and that requiring scipy 1.0 for lmfit 1.0 (maybe next release? or maybe after a 0.9.10?) would be fine.   

Maybe we should aim for a lmfit 0.9.10 that includes ampgo and basinhopping, and be approximately the final Python 2 release, and then have lmfit 1.0 require scipy 1.0 and Python 3.6?   Not a strong opinion, but lmfit 1.0 should be soon and "drop support for Python2" should be soon too, so maybe those should be the same release....

Do you agree?  Anyone else have any opinions?

reneeotten commented:
I would be in favor of releasing a 0.9.10 version that includes these PRs. It's pretty straightforward to get this to work in scipy version < 0.19 though and I think it's worthwhile doing so for now. I'll push the code that will take care of this and you can decide whether you like it or not.

I do agree that with a 1.0 release of lmfit one can certainly push the requirements and at that time requiring scipy => 1.0 and not testing explicitly with Python2 seems reasonable to me. Before that to happen though I think there are a few more things that ideally should be incorporated (e.g., estimating uncertainties for all minimizers, using the external ```uncertainties``` package, ....).

reneeotten commented:
@newville I was passing ```bounds``` to the underlying minimizer, which isn't very useful since most of the ```scipy.optimize.minimize``` minimizers do not support bounds anyway - which is kind of the point of ```lmfit``` ;) So that wasnt' too smart! Anyway, now I changed it so that is uses ```result.init_vals``` as the starting point and does do the ```apply_bounds_transformation``` when calculating the ```penalty```. That should make sure that for all minimizers the parameters stay within bounds, or...?

I think it is still necessary to have the ```accept_test``` present to figure out if the proposed step is within bounds. In that case, can I test directly check whether ```x_new``` is within bounds or should I do ```apply_bounds_transformation``` first?

I would appreciated if you could take a little closer look at the code and give your opinion on this as well as on the test for ```scipy``` < 0.19. Once these last issues are worked out, I think it should be ready to go!

newville commented:
@reneeotten  OK, thanks.  I will look at this with full attention next week.   From what I've seen so far, it looks like merging this and #466 should be fine, and then we'll be approximately ready for 0.9.10.

reneeotten commented:
@newville sounds good, let's try to get this done next week! I did look a bit further into the basinhopping code and am now rather convinced there is no need for the ```accept_test``` as this is done after the random step + minimization from that point.  The current code uses the "normal" lmfit transformations to stay within the bounds, which should work for all the solvers in ```scipy.optimize.minimize``` and will never allow a parameter to go outside of its bounds. The random step will be done from the "internally used, freely variable values" and then transformed again to the bounded version so also that step should never go outside the bounds.

In my (admittedly not exhaustive testing) this all seems to work as intended. I added the ```Alpine02``` function, which is used in the ```scipy``` global benchmark to the test and the implementation of basinhopping in ```scipy``` doesn't find the global minimum from these starting points (unless one increases the number of iterations), whereas the version in ``lmfit``` does with the default settings. I think this is a positive side-effect of always staying within the bounds, but it's difficult to do an exact comparison to the ```scipy``` version...

Either way, I will now not touch this branch anymore until you had a chance to take a closer look!

newville commented:
Sorry got caught up in another meeting.. I'd be happy to merge this, so if you have changes, go for it.

reneeotten commented:
@newville okay, I rebased with master and cleaned-up the commits a bit.

If you looked at the code and agree with the implementation (i.e., using the the apply_bounds_transformation, not supplying explicit bounds to the underlying optimizer, and such; see my longer comment from a week ago), then it should be ready to merge. I will then update the ampgo PR to use this same approach.

newville commented:
@reneeotten Thanks!  I *do* like the use of apply_bounds_transformation -- it seems more consistent with the rest of lmfit and better keeps the code for applying bounds in one place. I think handling bounds per-solver is OK, but if the internal bounds can be used that's even better.  If ampgo can do that too, that would be fantastic.

So, I'll happily merge -- this is all really great stuff!   Thanks again!
